doc/articles/StructChaining.md

```diff
@@ -0,0 +1,81 @@
+
+An extensible struct is statically typed. It is the root of a linked list, containing a pointer to the next struct in the chain:
+
+```c
+typedef struct WGPUMyStructBase {
+    WGPUChainedStruct * nextInChain;
+    uint32_t x;
+} WGPUMyBaseStruct;
+```
+
+Each extension struct is a "subtype" of @ref WGPUChainedStruct; that is, its first member is a @ref WGPUChainedStruct, so that it can be safely cast to @ref WGPUChainedStruct. This allows the implementation to read its @ref WGPUChainedStruct::sType, which is some value of @ref WGPUSType ("Struct Type") that dynamically identifies the struct's type. (The `sType` may come from an implementation-specific extension; @ref WGPUSType is an "open" enum.)
+
+Once the implementation identifies the struct by its `sType`, it casts the pointer back to the appropriate struct type in order to access its contents. Setting `sType` incorrectly (or pointing to any type that isn't a subtype of @ref WGPUChainedStruct) causes undefined behavior.
+
+```c
+typedef enum WGPUSType {
+    // ...
+    WGPUSType_MyStructExtension1 = /* ... */,
+    WGPUSType_MyStructExtension2 = /* ... */,
+    // ...
+} WGPUStype;
+
+typedef struct WGPUMyStructExtension1 {
+    WGPUChainedStruct chain; // .chain.sType must be WGPUSType_MyStructExtension1
+    uint32_t y;
+} WGPUMyStructExtension1;
+
+typedef struct WGPUMyStructExtension2 {
+    WGPUChainedStruct chain; // .chain.sType must be WGPUSType_MyStructExtension2
+    uint32_t z;
+} WGPUMyStructExtension2;
+```
+
+This is used like so:
+
+```c
+WGPUMyStructExtension2 ext2 = WGPU_MY_STRUCT_EXTENSION2_INIT;
+// .chain.sType is already set correctly by the INIT macro.
+// .chain.next is set to NULL indicating the end of the chain.
+ext2.z = 2;
+
+WGPUMyStructExtension1 ext1 = WGPU_MY_STRUCT_EXTENSION1_INIT;
+// .chain.sType is already set correctly by the INIT macro.
+// .chain.next may be set in either of two ways, equivalently:
+ext1.chain.next = &ext2.chain;
+ext1.chain.next = (WGPUChainedStruct*) &ext2;
+ext1.y = 1;
+
+WGPUMyStructBase base = WGPU_MY_STRUCT_BASE_INIT;
+// Note: base structs do not have STypes (they are statically typed).
+base.nextInChain = &ext1.chain;
+base.x = 0;
+```
+
+The pointer links in a struct chain are all mutable pointers. Whether the structs in the chain are actually mutated depends on the function they are passed to (whether the struct chain is passed as an input or an output). Some structs (e.g. @ref WGPULimits) may be either an input or an output depending on the function being called.
+
+## Struct-Chaining Error {#StructChainingError}
+
+A struct-chaining error occurs if a struct chain is incorrectly constructed (in a detectable way).
+They occur if and only if:
+
+- The `sType` of a struct in the chain is not valid for the type of the chain root.
+- Multiple instances of the same `sType` value are seen in the same struct chain. (Note this also detects and disallows cycles.)
```

Comment:
nit. To clarify this a bit more (at least w.r.t the fuzzer case), it's not just that the same `sType` can't be in the same chain, but that the same `sType` should never appear in a chain or member chain on a root struct. Not sure how to word it clearly here...

Reply:
(discussed offline)

The first requirement (sType is valid for the type of the chain root) should be the one that validates that. The only gap is if some implementation adds an extension that can _actually_ recurse, which we probably should not allow them to do, so I'll note that.

We should also guarantee that if the implementation doesn't allow the sType in that context, it will never try to downcast the struct. It's the same as if the sType were totally bogus.

In Dawn's wire we came up with a way to guarantee that invalid chains can be serialized and deserialized without hitting serialization errors: if the client doesn't allow an sType in context, convert it to something like `WGPUSType_DawnInvalidChainedStruct` which indicates `struct WGPUDawnInvalidChainedStruct { WGPUChainedStruct chain; WGPUSType sTypeForErrorMessage; }`. As long as the client and server are in sync the server can always handle this without a deserialization error, and forward it into dawn-native which will produce a validation error. If the client is out of sync or is malicious (or a fuzzer) then if the server sees an sType that's not valid in context it immediately gives up and fails deserialization.

Reply:
> The only gap is if some implementation adds an extension that can _actually_ recurse, which we probably should not allow them to do, so I'll note that.

Actually I started trying to write this down and realized there _could_ be legitimate use cases that pass variable-sized data as a linked list (rather than an array or array of pointers), regardless of whether they're using struct chaining. So I'm going to make this a "should", not a "must".